File: project2/Player.py
from Pente import *

import copy
import threading
import time

class Player:
  def __init__(self, timeLimit):
    self._timeLimit = timeLimit
    self._lock = threading.Lock()
    
    self._consideredMoves = []
    self._selectedMove = [-1,-1]
    self._horizon = PenteState()
    self._value = 0
    
  def startTimer(self):
    self._lock.acquire()
    self._startTime = time.time()
    self._lock.release()
    
  def turnOver(self):
    over = (time.time() - self._startTime > self._timeLimit)
    return over
    
  def selectMove(self, move):
    self._lock.acquire()
    self._selectedMove = move
    self._lock.release()
    
  def getSelectedMove(self):
    self._lock.acquire()
    move = copy.copy(self._selectedMove)
    self._lock.release()
    return move
    
  def clearConsideredMoves(self):
    self._lock.acquire()
    self._consideredMoves = []
    self._lock.release()
    
  def considerMove(self, move):
    self._lock.acquire()
    self._consideredMoves.append(move)
    self._lock.release()
    
  def getConsideredMoves(self):
    self._lock.acquire()
    considered = copy.copy(self._consideredMoves)
    self._lock.release()
    return considered
    
  def selectHorizon(self, h, value):
    self._lock.acquire()
    self._horizon = h
    self._value = value
    self._lock.release()
    
  def getHorizon(self):
    self._lock.acquire()
    h = self._horizon
    self._lock.release()
    return h
    
  def getValue(self):
    self._lock.acquire()
    v = copy.copy(self._value)
    self._lock.release()
    return v 

  def maxPlayerMove(self, state):
    pass

  def minPlayerMove(self, state):
    pass
    

File: project2/Player.cpp
#include "Player.h"

Player::Player(double timeLimit) : _timeLimit(timeLimit) { srand(time(NULL)); }

void Player::startTimer() {
  std::lock_guard<std::mutex> lock(_accessLock);
  _startTime = std::chrono::high_resolution_clock::now();
}

bool Player::turnOver() const {
  //  std::lock_guard<std::mutex> lock(_accessLock);
  auto currentTime = std::chrono::high_resolution_clock::now();
  double duration = std::chrono::duration_cast<std::chrono::seconds>(
      currentTime - _startTime).count();
  return duration > _timeLimit;
}

void Player::selectMove(Position m) {
  std::cout << "(best now " << (char)(m[1] + 'A') << (int)(m[0] + 1) << ") ";
  std::lock_guard<std::mutex> lock(_accessLock);
  _selectedMove = m;
}

Position Player::getSelectedMove() const {
  std::lock_guard<std::mutex> lock(_accessLock);
  return _selectedMove;
}

void Player::clearConsideredMoves() {
  std::cout << "\nSearch new depth tree.  Nodes considered: ";
  std::lock_guard<std::mutex> lock(_accessLock);
  _consideredMoves.clear();
}

void Player::considerMove(Position m) {
  std::cout << (char)(m[1] + 'A') << (int)(m[0] + 1) << " ";
  std::lock_guard<std::mutex> lock(_accessLock);
  _consideredMoves.push_back(m);
}

std::vector<Position> Player::getConsideredMoves() const {
  std::lock_guard<std::mutex> lock(_accessLock);
  return _consideredMoves;
}

void Player::selectHorizon(PenteState h, int v) {
  std::lock_guard<std::mutex> lock(_accessLock);
  _moveValue = v;
  _horizon = h;
}

PenteState Player::getHorizon() const {
  std::lock_guard<std::mutex> lock(_accessLock);
  return _horizon;
}

int Player::getValue() const {
  std::lock_guard<std::mutex> lock(_accessLock);
  return _moveValue;
}

void Player::maxPlayerMove(PenteState s) {}

void Player::minPlayerMove(PenteState s) {}
